---
title: |
  ![](UF.jpg){width=10%}

  AlphaSimR: Implementing a breeding pipeline
  
author: 
  - Marcio Resende, ^[professor, University of Florida, mresende@ufl.edu]
 
  - Marco Antonio Peixoto, ^[Post doc, University of Florida, deamorimpeixotom@ufl.edu]
  
output: 
  html_document:
    toc: true
    toc_float: FALSE
    theme: flatly
    highlight: haddock
bibliography: references.bib
---

```{=html}
<style type="text/css">
  body{
  font-size: 12pt;
}
</style>
```

## 1. Introduction

To start a set of simulations in the `AlphaSimR` package [@gaynor2021alphasimr], four steps must be implemented, as follows:

1. Simulate founder genomes/haplotypes.
2. Set global simulation parameters for a target trait/traits.
3. Model the breeding program.
4. Examine results by looking into population individuals' metrics.

In this vignette, we will cover a few useful functions from **AlphaSimR** that will help to model a the breeding program. In addition, we will compare three pipelines over a few generations of selection.

## 2. Breeding program

Imagine that we got hired by a breeding company. In your first day, your team introduced you to the pipeline below:

![](Breeding.png)

Then, your manager mentioned that the you will implement changes in the pipeline and they gave you two options:

1) Increase the number of crosses from 50 to 70 (20 more families);
2) Increase the number of doubled-haploid (DH) lines produced from 10 to 14 (400 more DH individuals).

As there are no budget for implementing both, your task will be to choose each strategy is the best and why.

So, in the next topics, and following the 4 steps before mentioned, we will implement the pipeline and the scenarios to leverage evidence for a data-drive  decision.

## Step 1: Founder Genome

We will start by loading the packages.


```{r}
# Loading packages
# install.packages("AlphaSimR") # Install if you do not have it
library(AlphaSimR)
library(ggplot2)

```



Let's start the simulation itself by creating the base genome.

```{r}
# Creating the founding genome
founderGenomes = runMacs(nInd = 100, # Number of individuals that compose the genome
                         nChr = 10,   # Number of chromosome pairs of the target species
                         segSites = 50,  # number of segregation sites
                         species = "MAIZE") # We can use the base parameters and we have available in the package MAIZE, WHEAT, CATTLE, and GENERIC.

# Object created
founderGenomes
```

## Step 2: Traits

With the founder genomes in perspective, we can proceed to add the characteristics of the target trait or traits to the simulation. AlphaSimR has a vignette that we recommend you take a look at [link](https://cran.r-project.org/web/packages/AlphaSimR/vignettes/traits.pdf). It gives, thoroughly, an explanation on the traits of interest and how to interpret its effects.

We will start with a trait with only additive effects. For such, we have the following parameters:


```{r}
# Global simulation parameters from founder genomes.
SP = SimParam$new(founderGenomes)

# Additive trait
SP$addTraitA(nQtlPerChr = 20, # Number of QTL per chromosome
             mean = 10, # Trait additive mean
             var = 10) # Trait additive variance or varA

SP$setVarE(varE=20) # trait heritability

# 
#             varA                 10
# h2 = ------------------- = ------------- = 0.33
#       varA + varE/nreps       10 + 20
#

# QTL effects (for the traits)
SP$traits[[1]]@addEff

# Generating base population
basePop = newPop(founderGenomes)

# Gen param
# genParam(basePop)

# Parents
Parents = basePop

```

The implementation of traits in **AlphaSimR** follows a biological model, which is responsible for converting into a genetic value each individual genotype before created. In a straightforward way, the genetic value is used to create the individuals' phenotypes. The biological effects presented in **AlphaSimR** are:

**A**: additive effect  
**D**: dominance effect  
**G**: genotype by environment interaction effect  
**E**: environmental effect  

So, we can create traits with the combinations of those effects (assuming that all of them as, at least, additive) using the **ADGE** framework, as it follows:

```{r, eval = FALSE}
# Traits that can be created in AlphaSimR:
SP$addTraitA()
SP$addTraitAD()
SP$addTraitADG()
SP$addTraitADEG()
SP$addTraitAG()
SP$addTraitAE()
SP$addTraitAEG()

```

It is important to have in mind that for dominance effects (**D**) we set the mean of the dominance degree (0-1) and variance, whereas for genotype by environment effect (**G**) and environmental effect (**E**) we have just to adjust the variance.



## Step 3: Deploying the pipeline

First, at the beginning of each breeding cycle, we generate new crosses. We can implement random crosses as well as make oriented crosses by giving the list of crosses to make.

```{r}

#------------- Year 1 
F1pop = randCross(pop = Parents, nCrosses = 50, nProgeny = 2) # Crossing block

#------------- Year 2
DHpop = makeDH(pop=F1pop, nDH=10) # Doubled Haploids

#------------- Year 3
rowStage = setPheno(pop=DHpop, h2 = 0.1, reps = 1) # h2 = 0.1 (visual selection)
PrelimTrial = selectWithinFam(rowStage, nInd = 3, use = "pheno") 

#-------------  Year 4
PrelimTrial = setPheno(PrelimTrial, varE = 20, reps = 1) # h2 = 10/(10+20/1) = 0.33
AdvancTrial = selectInd(PrelimTrial, nInd = 50, use = "pheno") # Preliminary Trial

#------------- Year 5
AdvancTrial = setPheno(AdvancTrial, varE = 20, reps = 5) # h2 = 10/(10+20/5) = 0.71
EliteTrial = selectInd(AdvancTrial, nInd = 20, use = "pheno") # Advanced Trial

#------------- Year 6
EliteTrial = setPheno(EliteTrial, varE = 20, reps = 20) # h2 = 10/(10+20/20) = 0.90
Variety = selectInd(EliteTrial, nInd = 2, use = "pheno") # Elite Trial

#------------- Year 7    
# Release varieties

```

## Step 4: Comparision

### Scenario 1


```{r}

# Allocate vectors
nYears = 20 + 1 # +1 to store starting generation
meanGAll = numeric(nYears)
varGAll = numeric(nYears)

# Save the starting values
meanGAll[1] = meanG(basePop)
varGAll[1] = varG(basePop)

# From the same population
Parents = basePop


# Loop
for (i in 2:nYears){
  
#------------- Year 1 
F1pop = randCross(pop = Parents, nCrosses = 50, nProgeny = 2) # Crossing block

#------------- Year 2
DHpop = makeDH(pop=F1pop, nDH = 10) # Doubled Haploids

#------------- Year 3
rowStage = setPheno(pop=DHpop, h2 = 0.1, reps = 1)
PrelimTrial = selectWithinFam(rowStage, nInd = 3, use = "pheno") 

#-------------  Year 4
PrelimTrial = setPheno(PrelimTrial, varE = 20, reps = 1)
AdvancTrial = selectInd(PrelimTrial, nInd = 50, use = "pheno") # Preliminary Trial

#------------- Year 5
AdvancTrial = setPheno(AdvancTrial, varE = 20, reps = 5)
EliteTrial = selectInd(AdvancTrial, nInd = 20, use = "pheno") # Advanced Trial

#------------- Year 6
EliteTrial = setPheno(EliteTrial, varE = 20, reps = 20)
Variety = selectInd(EliteTrial, nInd = 2, use = "pheno") # Elite Trial

#------------- Year 7    
# Release varieties

# Updating the parents
Parents = AdvancTrial

# Track performance
meanGAll[i] = meanG(Parents)
varGAll[i] = varG(Parents)

}

# Track performance
meanG_Scen1 = meanGAll
varG_Scen1 = varGAll


```

### Scenario 2

```{r}

# Allocate vectors
nYears = 20 + 1 # +1 to store starting generation
meanGAll = numeric(nYears)
varGAll = numeric(nYears)

# Save the starting values
meanGAll[1] = meanG(basePop)
varGAll[1] = varG(basePop)

# From the same population
Parents = basePop


# Loop
for (i in 2:nYears){
  
#------------- Year 1 
F1pop = randCross(pop = Parents, nCrosses = 70, nProgeny = 2) # Crossing block

#------------- Year 2
DHpop = makeDH(pop=F1pop, nDH=10) # Doubled Haploids

#------------- Year 3
rowStage = setPheno(pop=DHpop, h2 = 0.1, reps = 1)
PrelimTrial = selectWithinFam(rowStage, nInd = 3, use = "pheno") 

#-------------  Year 4
PrelimTrial = setPheno(PrelimTrial, varE = 20, reps = 1)
AdvancTrial = selectInd(PrelimTrial, nInd = 50, use = "pheno") # Preliminary Trial

#------------- Year 5
AdvancTrial = setPheno(AdvancTrial, varE = 20, reps = 5)
EliteTrial = selectInd(AdvancTrial, nInd = 20, use = "pheno") # Advanced Trial

#------------- Year 6
EliteTrial = setPheno(EliteTrial, varE = 20, reps = 20)
Variety = selectInd(EliteTrial, nInd = 2, use = "pheno") # Elite Trial

#------------- Year 7    
# Release varieties

# Updating the parents
Parents = AdvancTrial

# Track performance
meanGAll[i] = meanG(Parents)
varGAll[i] = varG(Parents)

}

# Track performance
meanG_Scen2 = meanGAll
varG_Scen2 = varGAll

    
```

### Scenario 3



```{r}

# Allocate vectors
nYears = 20 + 1 # +1 to store starting generation
meanGAll = numeric(nYears)
varGAll = numeric(nYears)

# Save the starting values
meanGAll[1] = meanG(basePop)
varGAll[1] = varG(basePop)

# From the same population
Parents = basePop


# Loop
for (i in 2:nYears){
  
#------------- Year 1 
F1pop = randCross(pop = Parents, nCrosses = 50, nProgeny = 2) # Crossing block

#------------- Year 2
DHpop = makeDH(pop=F1pop, nDH=14) # Doubled Haploids

#------------- Year 3
rowStage = setPheno(pop=DHpop, h2 = 0.1, reps = 1)
PrelimTrial = selectWithinFam(rowStage, nInd = 3, use = "pheno") 

#-------------  Year 4
PrelimTrial = setPheno(PrelimTrial, varE = 20, reps = 1)
AdvancTrial = selectInd(PrelimTrial, nInd = 50, use = "pheno") # Preliminary Trial

#------------- Year 5
AdvancTrial = setPheno(AdvancTrial, varE = 20, reps = 5)
EliteTrial = selectInd(AdvancTrial, nInd = 20, use = "pheno") # Advanced Trial

#------------- Year 6
EliteTrial = setPheno(EliteTrial, varE = 20, reps = 20)
Variety = selectInd(EliteTrial, nInd = 2, use = "pheno") # Elite Trial

#------------- Year 7    
# Release varieties

# Updating the parents
Parents = AdvancTrial

# Track performance
meanGAll[i] = meanG(Parents)
varGAll[i] = varG(Parents)

}

# Track performance
meanG_Scen3 = meanGAll
varG_Scen3 = varGAll
    
```


### Plotting

```{r}

# Plot mean of genetic values over time
meanRanges = range(c(meanG_Scen1, meanG_Scen2, meanG_Scen3))
plot(x = 1:nYears, y = meanG_Scen1, type = "l", col = "black", lwd = 3,
     xlab = "Years", ylab = "Genetic values", ylim = meanRanges)
lines(x = 1:nYears, y = meanG_Scen2, type = "l", col = "blue", lty = 2, lwd = 3)
lines(x = 1:nYears, y = meanG_Scen3, type = "l", col = "grey", lty = 2, lwd = 3)

legend(x = "topleft",  legend = c('Scen1','Scen2','Scen3'), title = "Scenario",
       lwd = 3, lty = c(1, 2,2), col = c("black", "blue", 'grey'), bty = "n")


```    





## References

::: {#refs}
:::



